chainer/utils/type_check.py

```diff
@@ -0,0 +1,281 @@
+import numpy
+
+from chainer import cuda
+
+
+class TypeInfo(object):
+
+    def __init__(self, name, index, shape, dtype):
+        self.index = index
+        name = '{0}[{1}]'.format(name, index)
+        self.name = name
+        self.shape = tuple(Shape(x, i, name) for i, x in enumerate(shape))
+        self.dtype = DtypeExpr(dtype, name)
+        self.ndim = Member(len(self.shape), name, 'ndim')
+
+    def is_none(self):
+        return self.dtype is None
+
+    def __str__(self):
+        return self.name
+
+
+class TypeInfoTuple(tuple):
+
+    def _set_name(self, name):
+        self.name = name
+
+    def size(self):
+        return Member(len(self), self.name, 'size')
+
+
+def get_types(data, is_grad):
+    assert(isinstance(data, tuple))
+
+    if is_grad:
+        name = 'out_types'
+    else:
+        name = 'in_types'
+
+    info = TypeInfoTuple(
+        get_type(name, i, x, is_grad) for i, x in enumerate(data))
+    info._set_name(name)
+    return info
+
+
+def get_type(name, index, array, accept_none):
+    if accept_none and array is None:
+        # case that gradient is not given
+        return TypeInfo(name, index, (), None)
+
+    assert(isinstance(array, numpy.ndarray) or
+           isinstance(array, cuda.GPUArray))
+    return TypeInfo(name, index, array.shape, array.dtype)
+
+
+def _wrap(x):
+    if isinstance(x, IntExpr):
+        return x
+    elif isinstance(x, int):
+        return IntConstant(x)
+    else:
+        assert(False)
+
+
+def _make_un_operator(exp, priority, func):
+    def f(x):
+        return UnaryOperator(_wrap(x), exp, priority, func)
+    return f
+
+
+def _make_bin_operator(exp, priority, func):
+    def f(x, y):
+        return BinaryOperator(_wrap(x), _wrap(y),
+                              exp, priority, func)
+    return f
+
+
+def _flip(f):
+    return lambda x, y: f(y, x)
+
+
+_add = _make_bin_operator('+', 4, int.__add__)
```

Comment:
Why these binary operators are outside of `IntExpr` and exposed? If it is not needed, I think it is better that they are in `IntExpr` class